src/nvim/terminal.c

```diff
@@ -651,6 +651,15 @@ static void buf_set_term_title(buf_T *buf, char *title)
   api_clear_error(&err);
 }
 
+// Get the associated buffer for the terminal and return whether it's in a
+// valid state to be used
+static buf_T *term_buf_valid(Terminal *term)
+  FUNC_ATTR_NONNULL_ALL
+{
+  buf_T *buf = handle_get_buffer(term->buf_handle);
```

Comment:
can't we make sure buffer is deleted from the handle dict at the same time as it is invalidated? Otherwise I would expect similar bug to sneak up sooner or later...

Reply:
Something like this?

```diff
diff --git a/src/nvim/buffer.c b/src/nvim/buffer.c
index b5ca6543c..b32c33869 100644
--- a/src/nvim/buffer.c
+++ b/src/nvim/buffer.c
@@ -459,6 +459,7 @@ void close_buffer(win_T *win, buf_T *buf, int action, int abort_if_last)
 
   if (buf->terminal) {
     terminal_close(buf->terminal, NULL);
+    handle_unregister_buffer(buf);
   }
 
   /* Always remove the buffer when there is no file name. */
```

`handle_unregister_buffer()` is already called by `close_buffer()` (via `free_buffer()`), but it happens a bit later in the function.  My only question here would be whether this introduces other issues because there are other parts of `close_buffer()` that expect the handle to still be registered.

Based on current test coverage, that doesn't seem to be the case, but that doesn't mean there isn't or that there won't be something added in the future which does need it.

Reply:
Adding more _ad-hoc_ logic like that certainly doesn't address my concern in any way, if anything it will make the code _harder_ to reason about... My point was that _invalid_ buffer handles have nothing to do in the buffer table, under _any condition_. Buffer handles are something introduces by nvim, I wouldn't expect vim's destruction code to rely on them in any way. 

>  that there won't be something added in the future which does need it.

And why would we add code that relies on invalid handles after we just removed them? Exactly what would be the point of code that relies on invalid memory?

Reply:
Well, it isn't invalid memory until it gets freed.  There are autocmds that run during `close_buffer`.  If any of those end up calling code that requires the handle, then they won't work anymore.

Reply:
I see, so it is necessary from a backwards compatibility point-of-view that arbitarary code will get exposed to a _partially invalidated_ buffer object...